### ISSUE TITLE
Level 6 

### ISSUE THREAD

**Issue by kaitlyngoh** [nus-cs2030/2122-s2#405]
hi can someone help on how to start for  level6? i read some posts about creating and additional pending event, but is it really necesssary? or can it be done without a pending event? if not, can someone explain how the pending event works? thanks!

**Comment by shotnothing** [comment]
There are definitely ways around creating this, but having a Pending/Rest/Resting event can make the later levels easier and cleaner to solve imo.

How it works is basically once there is a Done, instead of creating a Serve for the next customer, you create a Pending event (with time = Done time + Resting time from the supplier). This Pending event will take over the responsibility of creating the Serve event for the next customer in place of Done.

**Comment by bryankwe** [comment]
Hi, I am one of those who created an additional event. This resulted in me having to somewhat restart from Level 3 though.

This was what my execute() produced for each type of event:
Arrive --> Wait / Serve / Leave
Wait --> Try
Try --> Try / Serve
Serve --> Done

So basically for Try, you check whether the customer is first in the queue, if it is then return Serve, else return Try

Nevertheless, this is not the way and I'm sure there are more efficient ways available!

**Comment by kaitlyngoh** [comment]
> There are definitely ways around creating this, but having a Pending/Rest/Resting event can make the later levels easier and cleaner to solve imo.
> 
> How it works is basically once there is a Done, instead of creating a Serve for the next customer, you create a Pending event (with time = Done time + Resting time from the supplier). This Pending event will take over the responsibility of creating the Serve event for the next customer in place of Done.

hi i was thinking how about instead of creating a pending event after done, why not creat a serve event which takes in the done time + rest time? is that possible or would it complicate things? thanks !

**Comment by jamietan123** [comment]
I think if you serve immediately the problem would be getting the next customer in queue or if there aren't any customer then you need to return another event

**Comment by shotnothing** [comment]
> > There are definitely ways around creating this, but having a Pending/Rest/Resting event can make the later levels easier and cleaner to solve imo.
> > How it works is basically once there is a Done, instead of creating a Serve for the next customer, you create a Pending event (with time = Done time + Resting time from the supplier). This Pending event will take over the responsibility of creating the Serve event for the next customer in place of Done.
> 
> hi @shotnothing i was thinking how about instead of creating a pending event after done, why not creat a serve event which takes in the done time + rest time? is that possible or would it complicate things? thanks !

That's a good question! I actually started out not wanting to make an additional class, but there were a few quirks about the logic that they expect. For example:

**Set max waiting queue length = 2**

Server 1: [ 1 2 3 ]

Customer 1 done at Server 1. 
Remove 1 from the queue.
Create Serve event, for Customer 2, set in the future.

Server 1: [ 2 3 ]

Customer 4 arrives. 
Using common sense they should be able to queue in Server 1, but the logic they expect is that Customer 4 is unable to wait and leaves, I suspect because they expect Customer 1 to be dequeued only at the end of the resting period.

Of course, its mostly up to implementation and you could artificially increase the reporting size of the queue by 1 during the resting period.

I also remember a few other compelling reasons that made me create a new event but I can't exactly remember, its been a while. Nonetheless, can try with your method first. Adding code is usually easier than deleting code, should be easier to duct tape the resting event into your existing implementation later.

**Comment by shotnothing** [comment]
> I think if you serve immediately the problem would be getting the next customer in queue or if there aren't any customer then you need to return another event

You could just not add a new event  and process the existing done event.

**Comment by morksen123** [comment]
PQ<Customer> might work

**Comment by YangHaoying** [comment]
Hii! I was amongst those who added in the pending event class as well. My feeling is that it really depends on your previous logics and how much of it you’re willing to change. For me a new pending class can solve the problem of returning actual Serve event only for the first customer inside the queue while protecting other customers by returning some ‘fake’ events for them so they’re not lost and forgotten in my PQ. Hope this helps!

**Comment by Hamayah** [comment]
Yup i used PQ<Customer> and had my wait event run a pending event which decides whether to serve or push it to the next available server
